### DISCUSSION TITLE
Array of values from custom checkbox component

### DISCUSSION THREAD

**Discussion by lordvcs** [react-hook-form/react-hook-form#6250] [Q&A]
I know that its possible to get an array of values on submitting a form containing multiple inputs of type checkbox with the same name attribute like so: https://codesandbox.io/s/magical-bose-sjfin?file=/src/multi_checkbox.jsx
![image](https://user-images.githubusercontent.com/6374881/129692442-0fdd78b4-82b6-4002-b33e-4c6a83fe50fa.png)
and logs to console on submitting as follows which is expected:
![image](https://user-images.githubusercontent.com/6374881/129692522-3cd050e6-b93a-4480-8e74-0335a25ea966.png)

Now my question is how do I get the same behaviour in the case when my checkbox component is a custom component as follows:
https://codesandbox.io/s/magical-bose-sjfin?file=/src/multi_checkbox_custom.jsx
![image](https://user-images.githubusercontent.com/6374881/129692825-0d51167a-d480-47cd-a0e8-12205f37f7f6.png)
On submitting this form with all 3 checked, gives an unexpected output as follows:
![image](https://user-images.githubusercontent.com/6374881/129693010-b316d2ca-a7df-4430-a292-6420b4c3ee56.png)

How do  I get the output as an array, each containing the value of the respective checkbox as in the first example?

**Comment by bluebill1049** [comment]
build a single controlled component and have a local state hosted instead.

**Reply by lordvcs** [reply]
@bluebill1049 thanks for taking the time to reply. I did as your suggestion, however the output is still not an array. The following is what I get
![image](https://user-images.githubusercontent.com/6374881/129708622-f46246fb-504c-4caf-8ff8-70861b47d241.png)

This is the how the modified component looks like: 
https://codesandbox.io/s/magical-bose-sjfin?file=/src/multi_checkbox_custom.jsx
![image](https://user-images.githubusercontent.com/6374881/129708813-b13f45c1-6ce2-493d-884a-7a12b79e2da1.png)
Is this what you meant?

**Reply by bluebill1049** [reply]
your default state should be an array, and you should have a group of checkbox to map with a prop such as `options`, you can refer to some of the popular UI library such as chakra UI or MUI.

**Reply by lordvcs** [reply]
@bluebill1049 I am not sure I understood you completely, but this is how it looks like now:
https://codesandbox.io/s/magical-bose-sjfin?file=/src/multi_checkbox_custom.jsx
![image](https://user-images.githubusercontent.com/6374881/129723029-1c9ee64e-9b2a-47b6-8e43-90f595bc2dc3.png)
This still doesn't work in the intended way.

I know you are a busy man, is there any chance you could edit the sandbox the way you meant?

**Reply by lordvcs** [reply]
@bluebill1049 please can you help?

**Reply by bluebill1049** [reply]
can you build a component such as bloew

```tsx
const CheckBoxGroup = ({ options }) => {
  const [value, setValue] = React.useState([])

  return <div>{options.map(() => <input type="checkbox" onChange={() => {  setValue([...]) }} />)}</div> // update your local state so the value of your form value is an array of value
}


```

**Reply by lordvcs** [reply]
@bluebill1049  Is this what you meant?
https://codesandbox.io/s/magical-bose-sjfin?file=/src/new_multi_checkbox_custom.jsx
![image](https://user-images.githubusercontent.com/6374881/130049759-256a2c61-0ae6-4df6-aa5e-fa1781a25aa0.png)
This does get me an array of values but I could do this without using hooks form at all(or I did not quite understand what you meant). I dont see how I could get data on submitting the form in the handlesubmit callback or implementing validation and such.
Instead of doing it like the above I could easily do it like this which gives me the intended output:
https://codesandbox.io/s/magical-bose-sjfin?file=/src/multi_checkbox.jsx
![image](https://user-images.githubusercontent.com/6374881/130050083-34b675a7-3506-4ecb-905e-0c6a85655f4a.png)
The only reason I wanted each checkbox to be its own custom component is so that each of the custom component has its own style and UI structure to it without duplication of code.
Maybe I didnt understand you?


PS: thanks for creating this wonderful library and spending time on maintaining this

**Reply by what1s1ove** [reply]
Hey @lordvcs,I just wanted to create the same issue. But I'm glad I found yours 🙏
Hey @bluebill1049, I have the same problem and am not very clear why there are such big differences between register and useController. There are some best practices, how can I achieve the result as with a register, but using useController?

**Reply by lordvcs** [reply]
@what1s1ove exactly, I expected it to have similar behavior too.

**Reply by lordvcs** [reply]
@bluebill1049 hi, are there any plans in the future to implement such a behavior??
